### PR TITLE
Support for WS2812 RGB LED Strip v4.3

### PR DESCRIPTION
**Pull request by wizmo2**
Adds support for WS2812 addressable LEDs to display visualizations and LMS integration with Home-Assistant for custom alerts. This PR  follows up on the previous PR to master-cmake (#101) fixed and tested for esp-idf 4.3.2. 

Was originally based on the work @rochuck has done in the SqueezeAmpToo code (#46), but with significant changes.  

**_Current features_**
- Generic support for WS8212 based RGB strips upto 256 Leds.
- Built in Audio visualizer effects.
- Remote LED control through LMS CLI commands (for example Home-Assistant Notifications). 
- Full integration into LMS Plugin and new UI

[youtube link](https://youtu.be/1PFEcQugR4s)
[project link](https://github.com/wizmo2/TAudio-Case)

**_How To_**
1. For non-TAudio hardware, connect the D-IN to an unused gpio pin, and the GND to the ESP32 Gnd.  For larger strips you may (/will) need an external power supply for the VCC, otherwise you maybe able to use the 5V pin.

### CODE REVIEW

**Comment by wizmo2** [comment]
Tested on T-Audio and Esp32-CAM module (bluetooth out)

Hold cleared.  Turns out that the config breaks using the release version of 4.3.2.  Compiling with your docker fixed the issues.

**Comment by wizmo2** [comment]
I'm planning on making this a compile option (per cspot), and later adding a TAUDIO option to Kconfig.

@philippe44, would that resolve your concerns re flash/memory use?

**Comment by philippe44** [comment]
I've been able to find some room but I'm now close to have exhausted all options. Would you mind updating your PR to the latest commits so we can evaluate the additional size?

**Comment by wizmo2** [comment]
Will work on it next week.  Its not much

> I've been able to find some room but I'm now close to have exhausted all options. Would you mind updating your PR to the latest commits so we can evaluate the additional size?

I'll work on it next week.  Its not much

**Comment by sle118** [comment]
allright @wizmo2 I just changed the default branch to master-v4.3 just so you are on board with this.

**Comment by wizmo2** [comment]
@sle118 , @philippe44 

I merged the changes, but am stuck on creating test builds.  It's failing on generating 'nanopb'.  I get the same error on both the idfv4-master and the old idfv4.3 docker images. 

It was a while ago, but I had a similar problem on a previous build.  That time I found an option to not recompile the cspot codecs, but I'm damned if I can find it again!

Here are the logs
```
root@6eb73b50114b:/project# idf.py build
Executing action: all (aliases: build)
Running ninja in directory /project/build
Executing "ninja all"...
[2/806] Generating nanopb/generator/proto/nanopb_pb2.py
FAILED: esp-idf/spotify/cspot/nanopb/generator/proto/nanopb_pb2.py
cd /project/build/esp-idf/spotify/cspot && /project/components/spotify/cspot/bell/nanopb/generator/protoc -I/project/build/esp-idf/spotify/cspot/nanopb/generator/proto --python_out=/project/build/esp-idf/spotify/cspot/nanopb/generator/proto /project/build/esp-idf/spotify/cspot/nanopb/generator/proto/nanopb.proto
/bin/sh: 1: /project/components/spotify/cspot/bell/nanopb/generator/protoc: not found
[3/806] cd /project/build && /opt/esp/tools/cmake/3.16.4/bin/cmake -P /project/generate_debug_scripts.cmake
Generating debug script files
[5/806] Performing build step for 'bootloader'
ninja: no work to do.
ninja: build stopped: subcommand failed.
ninja failed with exit code 1
```

**Comment by wizmo2** [comment]
This is now merged and ready for review.

I've tested on three devices (SqueezeAmp soundbar, T-Audio line-out, and T-Watcher Bluetooth).  All seems OK.

**Comment by sle118** [comment]
@wizmo2 good work on bringing all the pieces together, including the command line and UI callback under the hardware tab.  You really grasped the concept there. 

One thing we typically don't do is to commit the sdkconfig alone, as we overwrite anything there from the build-script directory. If there's a config you absolutely need for this change to work, then you'd need to include that in the default sdkconfig if a config entry isn't in all the scripts, or change the scripts to make sure the value is correct for building targets.

That being said, I couldn't help but notice that you've increased the following
`CONFIG_FREERTOS_ISR_STACKSIZE=2096`
from 
`CONFIG_FREERTOS_ISR_STACKSIZE=1536`

since there's a ISR stack for each processor, it means `2x560=1120 bytes`  reserved right there even if there's no display or led strip. Is this required for the code to work? 

Since this is also touching @philippe44's realm, I will let him comment on the code review.

Thank you for the contribution, I'll definitely try it for fun!

**Comment by wizmo2** [comment]
Thanks.

Changes to `sdkconfig` are not needed (I left it in the PR because your build scripts overwrite any-how).  Not sure where the stacksize change came from, as I've been copy/pasting directly from the build-config folder.

**Comment by sle118** [comment]
I think your contribution is solid and ready for merging, but I will let @philippe44 take a final look before doing a Squash and merge. Thank you!

**Review by sle118 — APPROVED** [review]

**Review by sle118 — COMMENTED** [review]
Please try to revert the sdkconfig file to the current version in our 4.3 branch. I see you added it to the .gitignore. Don't... as we sometiemes have to push new defaults when, for example, switching ESP-IDF releases, fine tuning, etc.

**Review by sle118 — CHANGES_REQUESTED** [review]
Please make the requested changes

**Review by sle118 — APPROVED** [review]

**Comment by pintomax** [comment]
@sle118  I'm eager to get the new release with this new code, will it happen anytime soon?

File: components/services/accessors.c [hunk 1]
```diff
@@ -316,6 +316,28 @@ esp_err_t config_rotary_set(rotary_struct_t * config){
 	return err;	
 }
 
+/****************************************************************************************
+ * 
+ */
+esp_err_t config_ledvu_set(ledvu_struct_t * config){
+	int buffer_size=512;
+	esp_err_t err=ESP_OK;
+	char * config_buffer=malloc_init_external(buffer_size);
```
Comment (philippe44) [inline components/services/accessors.c:328]: Aren't we missing the free() here and why 2 config_buffer(s)?

**Comment by philippe44** [comment]
Also, there have been changes in Bell (Spotify) that should be ignored

**Comment by philippe44** [comment]
> @sle118 I'm eager to get the new release with this new code, will it happen anytime soon?
>
My bad, I've been away for a long while

**Comment by wizmo2** [comment]
@sle118, `config` / NVS seems to be broken in 1331 (idf 4.3.5 build).  

I made the change to accessors, but could not test as none of the NVS configurations worked.   I reverted to the release version of 1331 and had the same issue.  Downgraded to 1306 and everything was working again.  Going to have look to see if I can find the issue.

PS.  Will revert all the bell mods once it tests ok.  I'm still fighting to get cspot to build correctly without it!

**Comment by philippe44** [comment]
> @sle118, `config` / NVS seems to be broken in 1331 (idf 4.3.5 build).
> 
Weird, I'll look into that as I've not changed anything wrt NVS, AFAIR

**Comment by wizmo2** [comment]
@philippe44, did you get a chance to look at the web hardware configuration in the latest builds.  

Is it working for you? It may just be my build environment, but I have the same issue if I download a release version.
For example, if I make a change to the I2C Bus Parameters and press Save, the old system gives a responses and I see messages in the esp32 log.  The new builds just says 'Executing...', I see messages in the browser console, but nothing in the esp32 and the NVS is not updated.  I've tried various devices, including my SqueezeAmp.

I spent a while trying to figure out how wifi-manager and platform-console works, but I'm really not getting anywhere!  It appears that the 'commands,json' POST is being ignored.

_P.S.  The SPDIF Config display seems to have been broken for a while.  It does not show the active configuration in any of the versions._

**Comment by philippe44** [comment]
No I have not. It really @sle118's realm and not the part of the code that I know well.

**Comment by wizmo2** [comment]
@philippe44 ,  Thanks for 1336.  Have now been able to test (SqueezeAmp, T-Audio, and ESP32-CAM/SPDIF both 16 and 32bit).  Cspot changes has been reverted, so I think its good to go!  

_BTW, The LMS plugin is updated.  I created a new zip file, but the revision has not been bumped._

**Comment by philippe44** [comment]
Can you address my comment about xTaskCreate?

**Comment by wizmo2** [comment]
> xTaskCreate

Sorry, can you elaborate, not seeing any comments re xTask

**Comment by philippe44** [comment]
In my comment of your PR, line 316 of led_strip.c

**Comment by wizmo2** [comment]
> In my comment of your PR, line 316 of led_strip.c

See the code, don't see any comments.  `led_strip` was a sourced [library](https://github.com/Lucas-Bruder/ESP32_LED_STRIP), so I have not really looked at it.  Let me know what you'd like me to change and I'll do what I can.

**Comment by philippe44** [comment]
If you go on your PR and expand led_strip.c, it's strange you don't see anything line #316?

Anyway, we can't use xTaskCreate as it uses IRAM stack which is very low already. If you look around, I'm using xTaskCreateStatic everywhere with a stack explicitly in DRAM, preferably if the tasks always run as a static block of memory

**Comment by wizmo2** [comment]
let me know if that is what you were looking for.  Tests OK.

**Comment by philippe44** [comment]
I've merged it, done tiny changes and updated the plugin version. It's not a "release" tag yet. @wizmo2 , thanks for your contribution and patience...

The additional code size is just 6k which is very good but unfortunately we are losing another 1kB of internal RAM. I was expecting half of that for the TCB (around 400Bytes) but I can't figure out why it's more. It does not show up from just the mapping, it might be something more vicious that the linker does. Commenting-out the task only gain 500B, as expected so I really do not know. 

It's a minor problem here but this project is chewing a bit of internal memory every time and it's close to be not manageable. When using SPDIF an all features, we might be as low as 15-20kB, means we are actually out of memory

**Pull request merged**